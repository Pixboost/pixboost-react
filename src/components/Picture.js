import React, {Component} from 'react';
import PropTypes from 'prop-types';
import lozad from 'lozad';

class Picture extends Component {
  constructor(props) {
    super(props);
    this.pictureRef = React.createRef();
  }

  componentDidMount() {
    if (this.props.lazy) {
      lozad(this.pictureRef.current).observe();
    }
  }

  static bpSrc(config, bp) {
    if (bp.hide) {
      return 'data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==';
    }
    const op = bp.op || 'optimise';

    return `${config.domain.includes('//') ? '' : '//'}${config.domain}/api/2/img/${bp.src}/${op}${op.includes('?') ? '&' : '?'}auth=${config.apiKey}`;
  }

  render() {
    const {config, breakpoints, alt, lazy, ...rest} = this.props;
    let defaultBp;
    return (
      <picture {...rest} ref={this.pictureRef}>
        {Object.keys(breakpoints).map(b => {
          const bpConfig = config.breakpoints[b];
          const bp = breakpoints[b];

          if (!bpConfig) {
            console.warn(`pixboost-picture: Can't find breakpoint config for ${b}`);
            return;
          }
          if (!bpConfig.media) {
            defaultBp = bp;
            return;
          }

          return (
            <source key={b} media={bpConfig.media} srcset={Picture.bpSrc(config, bp)}/>
          );
        })}
        {defaultBp && !lazy &&
          <img src={Picture.bpSrc(config, defaultBp)} alt={alt}/>
        }
        {defaultBp && lazy &&
          <source srcset={Picture.bpSrc(config, defaultBp)} alt={alt}/>
        }
      </picture>
    )
  }
}

Picture.propTypes = {
  alt: PropTypes.string,
  lazy: PropTypes.bool,
  breakpoints: PropTypes.object.isRequired,
  config: PropTypes.object.isRequired
};

Picture.defaultProps = {
  lazy: true
};

export default Picture;