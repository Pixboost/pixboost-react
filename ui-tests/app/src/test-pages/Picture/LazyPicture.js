import React, { Component } from 'react';
import {pixboostConfig} from '../config';
import {Picture} from 'pixboost-react';

class LazyPicture extends Component {
  render() {
    return (
      <div>
        <div style={{minHeight: '2000px'}}/>
        <Picture config={pixboostConfig}
                 breakpoints={{
                   lg: {
                     src: 'http://www.midday.coffee/assets/cups/blur-2846257__340.jpg',
                   },
                   md: {
                     src: 'http://www.midday.coffee/assets/cups/code-geek-2680204__340.png',
                     op: 'fit?size=200x200'
                   },
                   sm: {
                     src: 'http://www.midday.coffee/assets/cups/coffee-3183729__340.jpg',
                     op: 'resize?size=x150'
                   }
                 }}/>
      </div>
    );
  }
}

export default LazyPicture;