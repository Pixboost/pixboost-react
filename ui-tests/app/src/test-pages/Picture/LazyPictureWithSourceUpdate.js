import React, { Component } from 'react';
import {Picture} from 'pixboost-react';
import {pixboostConfig} from '../config';

class LazyPictureWithSourceUpdate extends Component {
  constructor(props) {
    super(props);

    this.state = {
      src: 'http://www.midday.coffee/assets/cup.jpeg'
    }
  }

  render() {
    return (
      <div>
        <Picture config={pixboostConfig}
                 breakpoints={{
                   lg: {
                     src: this.state.src,
                   },
                   md: {
                     src: this.state.src,
                     op: 'fit?size=200x200'
                   },
                   sm: {
                     src: this.state.src,
                     op: 'fit?size=200x200'
                   }
                 }}/>
        <button onClick={() => {this.setState({src: 'http://www.midday.coffee/assets/banner.jpg'})}}>Update source</button>
      </div>
    );
  }
}

export default LazyPictureWithSourceUpdate;