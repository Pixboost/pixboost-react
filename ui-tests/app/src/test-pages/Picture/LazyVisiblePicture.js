import React, { Component } from 'react';
import {pixboostConfig} from '../config';
import {Picture} from 'pixboost-react';

class LazyVisiblePicture extends Component {
  render() {
    return (
      <Picture lazy={true}
               config={pixboostConfig}
               breakpoints={{
                 lg: {
                   src: 'http://www.midday.coffee/assets/cups/blur-2846257__340.jpg',
                 },
                 md: {
                   src: 'http://www.midday.coffee/assets/cups/code-geek-2680204__340.png',
                   op: 'fit?size=200x200'
                 },
                 sm: {
                   src: 'http://www.midday.coffee/assets/cups/coffee-3183729__340.jpg',
                   op: 'resize?size=x150'
                 }
               }}/>
    );
  }
}

export default LazyVisiblePicture;