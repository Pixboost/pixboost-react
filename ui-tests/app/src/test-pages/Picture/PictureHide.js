import React, { Component } from 'react';
import {pixboostConfig} from '../config';
import {Picture} from 'pixboost-react';

class PictureHide extends Component {
  render() {
    return (
      <React.Fragment>
        <span>You should not see an image on desktop and tablet.</span>
        <Picture lazy={true}
                 config={pixboostConfig}
                 breakpoints={{
                   lg: {
                     src: 'http://www.midday.coffee/assets/cups/blur-2846257__340.jpg',
                     hide: true
                   },
                   md: {
                     src: 'http://www.midday.coffee/assets/cups/code-geek-2680204__340.png',
                     hide: true
                   },
                   sm: {
                     src: 'http://www.midday.coffee/assets/cups/coffee-3183729__340.jpg',
                     op: 'resize?size=x150'
                   }
                 }}/>
      </React.Fragment>
    );
  }
}

export default PictureHide;