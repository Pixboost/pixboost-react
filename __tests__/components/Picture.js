import Picture from '../../src/components/Picture';
import React from 'react';
import renderer from 'react-test-renderer';

jest.mock('lozad', () => {
  return jest.fn().mockImplementation(() => {
    return {
      observe: () => {}
    };
  });
});

const testConfig = {
  domain: 'https://test.com',
  apiKey: 'abc123',
  breakpoints: {
    lg: {media: '(min-width: 990px)'},
    md: {media: '(min-width: 640px)'},
    sm: {}
  }
};

describe('Picture', () => {
  it('should render with minimum parameters', () => {
    expect(
      renderer.create(
        <Picture config={testConfig}
          breakpoints={{
            sm: {hide: true},
            md: {src: 'https://here.com/logo.png', op: 'resize?size=200'},
            lg: {src: 'https://here.com/logo-large.png'}
          }}
        />
      ).toJSON()
    ).toMatchSnapshot();
  });

  it('should render when not lazy', () => {
    expect(
      renderer.create(
        <Picture config={testConfig}
                 lazy={false}
                 breakpoints={{
                   sm: {hide: true},
                   md: {src: 'https://here.com/logo.png', op: 'resize?size=200'},
                   lg: {src: 'https://here.com/logo-large.png'}
                 }}
        />
      ).toJSON()
    ).toMatchSnapshot();
  });

  it('should render with alt attribute', () => {
    expect(
      renderer.create(
        <Picture config={testConfig}
                 alt={'YO!'}
                 breakpoints={{
                   sm: {hide: true},
                   md: {src: 'https://here.com/logo.png', op: 'resize?size=200'},
                   lg: {src: 'https://here.com/logo-large.png'}
                 }}
        />
      ).toJSON()
    ).toMatchSnapshot();
  });
});