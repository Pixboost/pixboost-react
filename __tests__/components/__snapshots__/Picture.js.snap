// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Picture should render when not lazy 1`] = `
<picture>
  <source
    media="(min-width: 640px)"
    srcset="https://test.com/api/2/img/https://here.com/logo.png/resize?size=200&auth=abc123"
  />
  <source
    media="(min-width: 990px)"
    srcset="https://test.com/api/2/img/https://here.com/logo-large.png/optimise?auth=abc123"
  />
  <img
    src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
  />
</picture>
`;

exports[`Picture should render with alt attribute 1`] = `
<picture>
  <source
    media="(min-width: 640px)"
    srcset="https://test.com/api/2/img/https://here.com/logo.png/resize?size=200&auth=abc123"
  />
  <source
    media="(min-width: 990px)"
    srcset="https://test.com/api/2/img/https://here.com/logo-large.png/optimise?auth=abc123"
  />
  <source
    alt="YO!"
    srcset="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
  />
</picture>
`;

exports[`Picture should render with minimum parameters 1`] = `
<picture>
  <source
    media="(min-width: 640px)"
    srcset="https://test.com/api/2/img/https://here.com/logo.png/resize?size=200&auth=abc123"
  />
  <source
    media="(min-width: 990px)"
    srcset="https://test.com/api/2/img/https://here.com/logo-large.png/optimise?auth=abc123"
  />
  <source
    srcset="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
  />
</picture>
`;
